README.md
# Hexdump2

An imperfect replica of `hexdump -C`.  API compatible with the [hexdump package's](https://pypi.org/project/hexdump/) hexdump function, with a few extra features.

## Installation

Use `pip`:
```commandline
pip install hexdump2   
```

## Usage

### As Python Library

Import the `hexdump` function from `hexdump2` package and use. By default, `hexdump` will internally output to stdout using the built-in `print` function:

```python
from hexdump2 import hexdump

hexdump(bytes(32))
"""
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000020
"""
```

The `hexdump` function has three ways to provide output, selected by the `result` keyword arg:
1. Print to stdout via the built-in print method - `print` (default)
2. Return a string with all the lines; useful for logging - `return`
3. Return a generator - `generator`

Additional functionality by keyword args:
* `offset` - specify an offset for the address. Default: 0
* `collapse` - turn on/off duplicate lines with `*`. Default: true
* `color` - turn on/off ANSI color codes (provided by [colorama](https://pypi.org/project/colorama/) package). Default: false

Color can be en/disabled all the time with by calling `color_always()` in python or by setting the environmental variable `HD2_EN_CLR` before importing.

```python
from hexdump2 import hexdump, color_always
color_always()  # Defaults to True
hexdump(bytes(32))
"""
[32m00000000[39m  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m|[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m|
[31m*
[32m00000020
"""
# Disable
color_always(False)
```

## Command Line

A simple command line is provided by running the python console script `hexdump2` or `hd2`:

```commandline
hexdump2 -h
```

If not installed via pip, the command line can also be run with:

```commandline
python3 path/to/hexdump2_package -h 
```

Color can be enabled all the time for the command line by setting the environmental variable `HD2_EN_CLR` with any value (internally it's a string, which is converted to a bool):

```commandline
export HD2_EN_CLR="True"
hd2 0x20_nulls.bin
[32m00000000[39m  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m|[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m|
[31m*
[32m00000020
```

## Usage Examples

Supply an offset:

```python
from hexdump2 import hexdump
   
hexdump(bytes(32), offset=0x100)
"""
00000100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000120
"""
```

Return a string for logging:

```python
import logging
from hexdump2 import hexdump

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("Example")
logger.setLevel(logging.INFO)

logger.info(
    f"Example data\n"
    f"{hexdump(bytes(32), result='return')}"
)
"""
INFO:Example:Example data
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000020
"""
```

Using the generator, where the `*` operator runs the generator into the `print` method to stderr:

```python
import sys
from hexdump2 import hexdump

print(
    *hexdump(range(256), result='generator'),
    file=sys.stderr
)
"""
00000000  00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f  |................|
00000010  10 11 12 13 14 15 16 17  18 19 1a 1b 1c 1d 1e 1f  |................|
00000020  20 21 22 23 24 25 26 27  28 29 2a 2b 2c 2d 2e 2f  | !"#$%&'()*+,-./|
00000030  30 31 32 33 34 35 36 37  38 39 3a 3b 3c 3d 3e 3f  |0123456789:;<=>?|
00000040  40 41 42 43 44 45 46 47  48 49 4a 4b 4c 4d 4e 4f  |@ABCDEFGHIJKLMNO|
00000050  50 51 52 53 54 55 56 57  58 59 5a 5b 5c 5d 5e 5f  |PQRSTUVWXYZ[\]^_|
00000060  60 61 62 63 64 65 66 67  68 69 6a 6b 6c 6d 6e 6f  |`abcdefghijklmno|
00000070  70 71 72 73 74 75 76 77  78 79 7a 7b 7c 7d 7e 7f  |pqrstuvwxyz{|}~.|
00000080  80 81 82 83 84 85 86 87  88 89 8a 8b 8c 8d 8e 8f  |................|
00000090  90 91 92 93 94 95 96 97  98 99 9a 9b 9c 9d 9e 9f  |................|
000000a0  a0 a1 a2 a3 a4 a5 a6 a7  a8 a9 aa ab ac ad ae af  |................|
000000b0  b0 b1 b2 b3 b4 b5 b6 b7  b8 b9 ba bb bc bd be bf  |................|
000000c0  c0 c1 c2 c3 c4 c5 c6 c7  c8 c9 ca cb cc cd ce cf  |................|
000000d0  d0 d1 d2 d3 d4 d5 d6 d7  d8 d9 da db dc dd de df  |................|
000000e0  e0 e1 e2 e3 e4 e5 e6 e7  e8 e9 ea eb ec ed ee ef  |................|
000000f0  f0 f1 f2 f3 f4 f5 f6 f7  f8 f9 fa fb fc fd fe ff  |................|
"""
```

### Non-Bytes Classes

Internally, hexdump will chunk an object by slicing.  The sliced object should support checking each item as an `int` (e.g., bytes(1)[0] == 0).

```python
import array
from hexdump2 import hexdump

data = array.array('B', bytes(16))
hexdump(data)
"""
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
"""
```

A class needs to implement `__bytes__`, `__len__`, and `__getitem__` to support hexdump:
```python
from hexdump2 import hexdump

class SomeStruct:
    def __init__(self, data: bytes):
        self._data = data
        
    def __bytes__(self):
        return self._data
    
    def __len__(self):
        return len(self._data)
    
    def __getitem__(self, item):
        return self._data.__getitem__(item)

some_struct = SomeStruct(bytes(32))
hexdump(some_struct)
"""
00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00000020
"""
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.1]
### Added

Support for NO_COLOR environmental variable (http://no-color.org/)
Support for checking of TERM environmental variable as "dumb"
Type hints for literal values possible to hexdump2 result kwarg.
Development install dependencies

### Fixed

Incorrect package importlib_metadata for Python>3.8
Incorrect link for changelog in setup.cfg metadata

### Changed

Run the black code formatter (https://black.readthedocs.io/en/stable/)

## [1.1.0] - 2022-02-8
### Added
- This changelog and URL for changelog into setup.cfg
- Requirements.txt for funsies

### Changed
- Command line switch `-a` for collapse changed to `-v` to match hexdump(1).
- Updated readme with information on setting color all the time.
- Updated hexdump2 package docstring

### Fixed
- Bug where a newline should be printed after a KeyboardInterrupt when running the CLI. 


tests/test_hexdump2.py
#!/usr/bin/env python3
import array
import contextlib
import importlib
import os
import sys
import tempfile
import unittest
from io import StringIO
from os import environ, linesep
from types import GeneratorType
from unittest.mock import patch

from hexdump2 import hd, hexdump
from hexdump2.__main__ import main

single_line_result = (
    f"00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
    f"00000010"
)
double_line_result = (
    f"00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
    f"*{linesep}"
    f"00000020"
)

nine_range_result = (
    "00000000  00 01 02 03 04 05 06 07  08                       |.........|\n00000009"
)

nine_range_color_result = r"""[32m00000000  [39m00 [36m01 [36m02 [36m03 [36m04 [36m05 [36m06 [36m07  [36m08                       [39m|[39m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m00000009[39m"""

nine_bytes_result = (
    "00000000  00 00 00 00 00 00 00 00  00                       |.........|\n00000009"
)

range_0x100_result = r"""00000000  00 01 02 03 04 05 06 07  08 09 0a 0b 0c 0d 0e 0f  |................|
00000010  10 11 12 13 14 15 16 17  18 19 1a 1b 1c 1d 1e 1f  |................|
00000020  20 21 22 23 24 25 26 27  28 29 2a 2b 2c 2d 2e 2f  | !"#$%&'()*+,-./|
00000030  30 31 32 33 34 35 36 37  38 39 3a 3b 3c 3d 3e 3f  |0123456789:;<=>?|
00000040  40 41 42 43 44 45 46 47  48 49 4a 4b 4c 4d 4e 4f  |@ABCDEFGHIJKLMNO|
00000050  50 51 52 53 54 55 56 57  58 59 5a 5b 5c 5d 5e 5f  |PQRSTUVWXYZ[\]^_|
00000060  60 61 62 63 64 65 66 67  68 69 6a 6b 6c 6d 6e 6f  |`abcdefghijklmno|
00000070  70 71 72 73 74 75 76 77  78 79 7a 7b 7c 7d 7e 7f  |pqrstuvwxyz{|}~.|
00000080  80 81 82 83 84 85 86 87  88 89 8a 8b 8c 8d 8e 8f  |................|
00000090  90 91 92 93 94 95 96 97  98 99 9a 9b 9c 9d 9e 9f  |................|
000000a0  a0 a1 a2 a3 a4 a5 a6 a7  a8 a9 aa ab ac ad ae af  |................|
000000b0  b0 b1 b2 b3 b4 b5 b6 b7  b8 b9 ba bb bc bd be bf  |................|
000000c0  c0 c1 c2 c3 c4 c5 c6 c7  c8 c9 ca cb cc cd ce cf  |................|
000000d0  d0 d1 d2 d3 d4 d5 d6 d7  d8 d9 da db dc dd de df  |................|
000000e0  e0 e1 e2 e3 e4 e5 e6 e7  e8 e9 ea eb ec ed ee ef  |................|
000000f0  f0 f1 f2 f3 f4 f5 f6 f7  f8 f9 fa fb fc fd fe ff  |................|
00000100"""

colored_ascii_range = r"""[32m00000000  [39m00 [36m01 [36m02 [36m03 [36m04 [36m05 [36m06 [36m07  [36m08 [36m09 [36m0a [36m0b [36m0c [36m0d [36m0e [36m0f  [39m|[39m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m00000010  [36m10 [36m11 [36m12 [36m13 [36m14 [36m15 [36m16 [36m17  [36m18 [36m19 [36m1a [36m1b [36m1c [36m1d [36m1e [36m1f  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m00000020  [33m20 [33m21 [33m22 [33m23 [33m24 [33m25 [33m26 [33m27  [33m28 [33m29 [33m2a [33m2b [33m2c [33m2d [33m2e [33m2f  [39m|[33m [33m![33m"[33m#[33m$[33m%[33m&[33m'[33m([33m)[33m*[33m+[33m,[33m-[33m.[33m/[39m|
[32m00000030  [33m30 [33m31 [33m32 [33m33 [33m34 [33m35 [33m36 [33m37  [33m38 [33m39 [33m3a [33m3b [33m3c [33m3d [33m3e [33m3f  [39m|[33m0[33m1[33m2[33m3[33m4[33m5[33m6[33m7[33m8[33m9[33m:[33m;[33m<[33m=[33m>[33m?[39m|
[32m00000040  [33m40 [33m41 [33m42 [33m43 [33m44 [33m45 [33m46 [33m47  [33m48 [33m49 [33m4a [33m4b [33m4c [33m4d [33m4e [33m4f  [39m|[33m@[33mA[33mB[33mC[33mD[33mE[33mF[33mG[33mH[33mI[33mJ[33mK[33mL[33mM[33mN[33mO[39m|
[32m00000050  [33m50 [33m51 [33m52 [33m53 [33m54 [33m55 [33m56 [33m57  [33m58 [33m59 [33m5a [33m5b [33m5c [33m5d [33m5e [33m5f  [39m|[33mP[33mQ[33mR[33mS[33mT[33mU[33mV[33mW[33mX[33mY[33mZ[33m[[33m\[33m][33m^[33m_[39m|
[32m00000060  [33m60 [33m61 [33m62 [33m63 [33m64 [33m65 [33m66 [33m67  [33m68 [33m69 [33m6a [33m6b [33m6c [33m6d [33m6e [33m6f  [39m|[33m`[33ma[33mb[33mc[33md[33me[33mf[33mg[33mh[33mi[33mj[33mk[33ml[33mm[33mn[33mo[39m|
[32m00000070  [33m70 [33m71 [33m72 [33m73 [33m74 [33m75 [33m76 [33m77  [33m78 [33m79 [33m7a [33m7b [33m7c [33m7d [33m7e [36m7f  [39m|[33mp[33mq[33mr[33ms[33mt[33mu[33mv[33mw[33mx[33my[33mz[33m{[33m|[33m}[33m~[36m.[39m|
[32m00000080  [36m80 [36m81 [36m82 [36m83 [36m84 [36m85 [36m86 [36m87  [36m88 [36m89 [36m8a [36m8b [36m8c [36m8d [36m8e [36m8f  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m00000090  [36m90 [36m91 [36m92 [36m93 [36m94 [36m95 [36m96 [36m97  [36m98 [36m99 [36m9a [36m9b [36m9c [36m9d [36m9e [36m9f  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000a0  [36ma0 [36ma1 [36ma2 [36ma3 [36ma4 [36ma5 [36ma6 [36ma7  [36ma8 [36ma9 [36maa [36mab [36mac [36mad [36mae [36maf  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000b0  [36mb0 [36mb1 [36mb2 [36mb3 [36mb4 [36mb5 [36mb6 [36mb7  [36mb8 [36mb9 [36mba [36mbb [36mbc [36mbd [36mbe [36mbf  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000c0  [36mc0 [36mc1 [36mc2 [36mc3 [36mc4 [36mc5 [36mc6 [36mc7  [36mc8 [36mc9 [36mca [36mcb [36mcc [36mcd [36mce [36mcf  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000d0  [36md0 [36md1 [36md2 [36md3 [36md4 [36md5 [36md6 [36md7  [36md8 [36md9 [36mda [36mdb [36mdc [36mdd [36mde [36mdf  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000e0  [36me0 [36me1 [36me2 [36me3 [36me4 [36me5 [36me6 [36me7  [36me8 [36me9 [36mea [36meb [36mec [36med [36mee [36mef  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m000000f0  [36mf0 [36mf1 [36mf2 [36mf3 [36mf4 [36mf5 [36mf6 [36mf7  [36mf8 [36mf9 [36mfa [36mfb [36mfc [36mfd [36mfe [36mff  [39m|[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[36m.[39m|
[32m00000100[39m"""

colored_0x100_nulls = r"""[32m00000000  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00 [39m00  [39m|[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m.[39m|
[31m*
[32m00000100[39m"""


class TestHexdump2(unittest.TestCase):
    def test_bad_return(self):
        with self.assertRaises(ValueError):
            hexdump(bytes(16), "bad_arg")

    def test_return_print(self):
        data = bytes(16)
        with StringIO() as buf, contextlib.redirect_stdout(buf):
            hexdump(data)
            buf.seek(0)
            r = buf.read()
            self.assertEqual(single_line_result + linesep, r)

    def test_return_print_newline_at_end(self):
        data = bytes(16)
        with StringIO() as buf, contextlib.redirect_stdout(buf):
            hexdump(data)
            print("Hello")

            buf.seek(0)
            r = buf.read()
            self.assertEqual(single_line_result + linesep + f"Hello{linesep}", r)

    def test_return_generator(self):
        data = bytes(16)
        r = hexdump(data, result="generator")

        self.assertTrue(isinstance(r, GeneratorType))
        self.assertEqual(single_line_result[:-8], next(r))

    def test_return_string(self):
        data = bytes(16)
        r = hexdump(data, "return")
        self.assertEqual(single_line_result, r)

    def test_address_offset(self):
        data = bytes(16)
        r = hexdump(data, "return", offset=0x100)
        self.assertEqual(
            f"00000100  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            f"00000110",
            r,
        )

    def test_short_line(self):
        self.assertEqual(nine_bytes_result, hexdump(bytes(9), "return"))

    def test_short_line_range(self):
        self.assertEqual(nine_range_result, hexdump(range(9), result="return"))

    def test_short_line_range_color(self):
        self.assertEqual(
            nine_range_color_result, hexdump(range(9), color=True, result="return")
        )

    def test_large_address(self):
        data = bytes(16)
        r = hexdump(data, "return", offset=(1 << 48) - 1)
        # noinspection SpellCheckingInspection
        self.assertEqual(
            f"ffffffffffff  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            f"100000000000f",
            r,
        )

    def test_multi_line(self):
        data = bytes(32)
        r = hexdump(data, "return", collapse=False)
        self.assertEqual(
            f"00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            f"00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            f"00000020",
            r,
        )

    def test_multi_line_no_nulls(self):
        r = hexdump(range(0x100), "return")
        self.assertEqual(range_0x100_result, r)

    def test_collapse_line(self):
        data = bytes(0x400)
        r = hexdump(data, result="return", collapse=True)
        self.assertEqual(
            f"00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            + f"*{linesep}"
            f"{len(data):08x}",
            r,
        )

    def test_no_data(self):
        data = b""
        r = hexdump(data, result="return")
        self.assertEqual("", r)

        with self.assertRaises(StopIteration):
            r = next(hexdump(data, result="generator"))
            self.assertEqual("", r)

    def test_large_data(self):
        data = bytes(2**25)
        r = hexdump(data, result="return")
        self.assertEqual(
            f"00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|{linesep}"
            + f"*{linesep}"
            f"{len(data):08x}",
            r,
        )

    def test_different_input_types(self):
        datas = (
            bytes(16),
            bytearray(16),
            array.array("B", bytes(16)),
        )

        for data in datas:
            with self.subTest():
                r = hexdump(data, result="return")
                self.assertEqual(r, single_line_result)

    def test_offset_no_data(self):
        r = hexdump(b"", result="return", offset=0x20)
        self.assertEqual(r, f"{0x20:08x}{linesep}")

    def test_color_all_colors(self):
        r = hexdump(range(0x100), result="return", color=True)
        self.assertEqual(colored_ascii_range, r)

    def test_color_collapse(self):
        r = hexdump(bytes(0x100), color=True, result="return")
        self.assertEqual(colored_0x100_nulls, r)

    @unittest.skip("Messes with other tests.")
    def test_color_always_env(self):
        current_env = environ.copy()

        current_env.setdefault("HD2_EN_CLR", "True")
        with patch.object(os, "environ", current_env):
            # Import so we can reload
            import hexdump2.hexdump2

            importlib.reload(hexdump2.hexdump2)

            r = hexdump2.hexdump2.hexdump(bytes(0x100), result="return")
            self.assertEqual(colored_0x100_nulls, r)

        # Reset the COLOR_ALWAYS flag
        importlib.reload(hexdump2.hexdump2)

    @unittest.skip("Messes with other tests.")
    def test_enable_color_always(self):
        from hexdump2.hexdump2 import color_always

        r = hexdump(bytes(0x100), result="return")
        self.assertNotEqual(colored_0x100_nulls, r)

        color_always()
        r = hexdump(bytes(0x100), result="return")
        self.assertEqual(colored_0x100_nulls, r)

        color_always(False)
        r = hexdump(bytes(0x100), result="return")
        self.assertNotEqual(colored_0x100_nulls, r)

    # noinspection PyTypeChecker
    def test_object_without_getitem(self):
        class Item:
            def __len__(self):
                # len of 0 will not raise as the code checks for zero length input
                return 1

        with self.assertRaises(TypeError) as cm:
            hexdump(Item())

        self.assertIn("subscriptable", cm.exception.args[0])

    # noinspection PyTypeChecker
    def test_object_without_len(self):
        class Item:
            pass

        with self.assertRaises(TypeError) as cm:
            hexdump(Item())

        self.assertIn("len()", cm.exception.args[0])

    def test_string_conversion(self):
        test_str = "".join(chr(_) for _ in range(256))
        r = hexdump(test_str, result="return")
        self.assertEqual(r, range_0x100_result)


class TestClassHD(unittest.TestCase):
    def test_in_string(self):
        data = bytes(32)
        with StringIO() as buf, contextlib.redirect_stdout(buf):
            r = f"{hd(data)}"
            self.assertEqual(double_line_result, r)

            buf.seek(0)
            r = buf.read()
            self.assertEqual("", r)

    def test_generator(self):
        data = bytes(32)
        with StringIO() as buf, contextlib.redirect_stdout(buf):
            for line in hd(data):
                print(line)

            buf.seek(0)
            r = buf.read()
            self.assertEqual(double_line_result + linesep, r)

    def test_print_in_script(self):
        with StringIO() as buf, contextlib.redirect_stdout(buf):
            print(hd(bytes(32)))

            buf.seek(0)
            r = buf.read()
            # We're not in an interactive script, so output should be empty
            self.assertEqual(double_line_result + linesep, r)


class TestCommandLineInterface(unittest.TestCase):
    def _call_main(self, expected_return_code: int = 0):
        with self.assertRaises(SystemExit) as cm:
            main()
        rc = cm.exception
        self.assertEqual(expected_return_code, rc.code)

    def test_no_args(self):
        with StringIO() as buf, contextlib.redirect_stderr(buf):
            self._call_main(2)

            buf.seek(0)
            r = buf.read()
            self.assertIn("usage: hexdump [-h]", r)

    def test_one_file(self):
        with tempfile.NamedTemporaryFile() as fh:
            fh.write(bytes(16))
            fh.seek(0)

            test_args = ["hexdump", fh.name]
            with patch.object(
                sys, "argv", test_args
            ), StringIO() as buf, contextlib.redirect_stdout(buf):
                self._call_main()

                buf.seek(0)
                r = buf.read()
                self.assertEqual(single_line_result + linesep, r)

    def test_multiple_files(self):
        num_test_files = 3
        files = [tempfile.NamedTemporaryFile() for _ in range(num_test_files)]
        for fp in files:
            fp.write(bytes(16))
            fp.seek(0)

        test_args = ["hexdump"] + [_.name for _ in files]
        with patch.object(
            sys, "argv", test_args
        ), StringIO() as buf, contextlib.redirect_stdout(buf):
            self._call_main()

            buf.seek(0)
            r = buf.read()
            self.assertEqual((single_line_result + linesep) * num_test_files, r)

        for fp in files:
            fp.close()

    def test_bad_auto_int(self):
        with tempfile.NamedTemporaryFile() as fh:
            fh.write(bytes(16))
            fh.seek(0)

            test_args = ["hexdump", fh.name, "-n", "abc"]
            with patch.object(
                sys, "argv", test_args
            ), StringIO() as buf, contextlib.redirect_stderr(buf):
                self._call_main(2)

    def test_no_colorama(self):
        import colorama
        import hexdump2.hexdump2

        old_colorama = colorama
        sys.modules["colorama"] = None
        importlib.reload(hexdump2.hexdump2)

        data = bytes(range(256))
        r = hexdump(data, "return", color=True)
        self.assertNotEqual(colored_ascii_range, r)

        sys.modules["colorama"] = old_colorama
        importlib.reload(hexdump2.hexdump2)


if __name__ == "__main__":
    unittest.main()


src/hexdump2/__main__.py
"""
Main function module for hexdump2
"""
import argparse
import sys
from os import linesep
from pathlib import Path

try:
    # Python 3.8+ should have this module
    from importlib.metadata import version
except ModuleNotFoundError:
    # For Python 3.6 and 3.7
    from importlib_metadata import version

from hexdump2 import hexdump


def _setup_arg_parser():
    """Creates argument parser for main function.

    :return: argpase arugment parser object
    """

    def _auto_int(value):
        try:
            return int(value, 0)
        except ValueError as e_auto_int:
            raise argparse.ArgumentTypeError(
                f"input cannot be converted to int type: {value}"
            ) from e_auto_int

    parser = argparse.ArgumentParser(
        prog="hexdump",
        description="An imperfect replica of hexdump -C",
    )
    parser.add_argument(
        "--version", action="version", version=f"%(prog)s {version('hexdump2')}"
    )
    parser.add_argument(
        "-n",
        dest="length",
        help="Interpret only length bytes of input.",
        type=_auto_int,
        required=False,
    )
    parser.add_argument(
        "-s",
        dest="offset",
        help="Skip offset bytes from the beginning of the input.",
        type=_auto_int,
        required=False,
    )
    parser.add_argument(
        "-l",
        dest="color",
        help="Color output",
        action="store_true",
    )
    parser.add_argument(
        "-v",
        dest="verbose_output",
        help="Cause Hexdump2 to display all lines; otherwise similar lines are shown with a `*`",
        action="store_false",
    )
    parser.add_argument(
        "file",
        help="One or more files to process. Each output is separated by a newline.",
        type=Path,
        nargs="+",
    )

    return parser


def main():
    """Main function run by console script hexdump2 or hd2.  Also run by `python -m hexdump2` on
    command line.
    :return: system exit return code.
    """
    parser = _setup_arg_parser()
    args = parser.parse_args()

    try:
        for file in args.file:
            with file.open("rb") as file_obj:
                file_size = file.lstat().st_size
                read_start = args.offset if args.offset is not None else 0
                length = args.length if args.length is not None else file_size

                hexdump(
                    file_obj.read(length),
                    offset=read_start,
                    collapse=args.verbose_output,
                    color=args.color,
                )
    except KeyboardInterrupt:
        # Caught interrupt; print a newline to make sure we're clear.
        print("")

    sys.exit(0)


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        raise SystemError(f"Caught exception in main:{linesep}{e}") from e


src/hexdump2/hexdump2.py
"""
Contains the functionality for creating hexdump lines from input data.
"""
from os import environ, linesep, name as os_name
from typing import ByteString, Iterator, Literal, Union

try:
    import colorama

    if os_name == "nt":
        colorama.init(autoreset=True)  # Only needed for Windows
except ImportError:
    colorama = None

COLOR_ALWAYS = bool(environ.get("HD2_EN_CLR", False))


def color_always(enable: bool = True):
    """Allows user to set flag to enable always coloring output
    :param enable: Always enable color
    """
    global COLOR_ALWAYS  # pylint: disable=global-statement

    # Support http://no-color.org/
    no_color = bool(environ.get("NO_COLOR", False))
    # Token check for color support
    dumb_terminal = environ.get("TERM")
    if no_color or dumb_terminal == "dumb":
        COLOR_ALWAYS = False
    else:
        COLOR_ALWAYS = enable


def _line_gen(
    data: ByteString, offset: int = 0x0, collapse: bool = True, color: bool = False
) -> Iterator[str]:
    """Generator function that yields a line.

    :param data: input data, must be bytes-like
    :param offset: offset for address
    :param collapse: flag to turn on/off collapsing multiple same lines
    :param color: enable color output; should only be used when outputting to stdout
    :return:
    """
    # Set color; colorama will import as None if not installed.
    if color and colorama:
        # address area
        address_color = colorama.Fore.GREEN
        # hex and ascii area
        zero_hex_color = colorama.Fore.RESET
        printable_color = colorama.Fore.YELLOW
        non_printable = colorama.Fore.CYAN
        # others
        star_line_color = colorama.Fore.RED

        # Cannot be set
        reset_color = colorama.Fore.RESET

    else:
        star_line_color = ""
        zero_hex_color = ""
        printable_color = ""
        non_printable = ""
        reset_color = ""
        address_color = ""

    # Empty data begets empty line
    if len(data) == 0:
        if offset:
            # Manifests when we've read past the end of a file, which results in an empty
            # buffer.
            # However, the offset we're reading at is still there.  Show the end address in this
            # case.
            # e.g., $ hexdump -s 10m 8mib_file.bin
            # 0800000
            yield f"{address_color}{offset:08x}{linesep}"

        # Return; this will cause a StopIteration
        return

    # Some sequences don't slice nicely (e.g. array.array('I', bytes(16));
    # test if we should convert to bytes
    if isinstance(data, (bytes, bytearray)):
        convert_to_bytes = False
    else:
        convert_to_bytes = True

    def _lookahead_gen():
        last_line_data = None
        yield_star = True

        for i in range(0, len(data), 16):
            line_data = data[i : i + 16]
            if collapse and line_data == last_line_data:
                # Only show the star once
                if yield_star:
                    yield_star = False
                    yield f"{star_line_color}*{linesep}"
                else:
                    # Otherwise, just goto the next data
                    continue
            else:
                if convert_to_bytes:
                    if isinstance(line_data, str):
                        # Use the `iso-8859-1` or `latin-1` encodings to map 0x00 to 0xff to bytes
                        # 0x00 to 0xff.
                        # c.f. https://docs.python.org/3/library/codecs.html#encodings-and-unicode
                        line_data = bytes(line_data, encoding="iso-8859-1")
                    else:
                        line_data = bytes(line_data)

                # address
                address_value = i + offset

                hex_str = ""
                ascii_str = ""
                hex_str_pad = 50  # 3 chr per position + 2 end spacing
                for j, byte in enumerate(line_data):
                    if byte == 0:
                        character_color = zero_hex_color
                        ascii_chr = "."

                    elif 0x20 <= byte <= 0x7E:
                        character_color = printable_color
                        ascii_chr = chr(byte)

                    else:
                        character_color = non_printable
                        ascii_chr = "."

                    ascii_str += f"{character_color}{ascii_chr}"
                    # Add double spaces after 8 bytes
                    if j != 7:
                        hex_str += f"{character_color}{byte:02x} "
                    else:
                        hex_str += f"{character_color}{byte:02x}  "

                    # Account for padding in color mode
                    hex_str_pad += len(character_color)

                yield_star = True
                yield f"{address_color}{address_value:08x}  {hex_str: <{hex_str_pad}}{reset_color}|{ascii_str}{reset_color}|{linesep}"

            last_line_data = line_data

    # Create a lookahead generator, this supports finding the last line, which we might need
    # to print the last address
    gen = _lookahead_gen()
    try:
        last = next(gen)
    except StopIteration:
        # Should never get here, as we check for empty data a head of time.  But...
        if offset:
            # if we have an offset, show it.
            yield f"{address_color}{offset:08x}{linesep}"

        # Return; this will cause a StopIteration
        return

    for line in gen:
        yield last
        last = line

    yield last
    # The last line; assume that receiver is using a function that will add a line seperator.
    yield f"{address_color}{len(data) + offset:08x}{reset_color}"


def hexdump(
    data: Union[ByteString, range],
    result: Literal["print", "return", "generator"] = "print",
    offset: int = 0x0,
    collapse: bool = True,
    color: bool = False,
) -> Union[None, str, Iterator[str]]:
    """Function that'll create `hexdump -C` of input data.

    :param data: bytes-like data to hexdump
    :param result: type of output/return, must be `print` (default), `return`, `generator`.
    :param offset: value to add to the address line.
    :param collapse: flag to turn on/off collapsing multiple same lines
    :param color: enable color output; should only be used for outputting to stdout
        (e.g. `result=print`).
    :return:
    """
    if COLOR_ALWAYS:
        color = COLOR_ALWAYS

    gen = _line_gen(data, offset, collapse, color)
    if result == "print":
        for line in gen:
            print(line, end="")

        # Add newline for last item
        print("")
        return None

    if result == "return":
        return "".join(gen)

    if result == "generator":
        return gen

    raise ValueError("`result` argument should be `print`, `return`, or `generator`")


# noinspection PyPep8Naming
class hd:
    """An attempt to get away from the `result` arg by caching the eventual string result.

    Calling this class in an interactive interpreter will cause the generator to be run and printed.
    Calling this within a string will also cause the generator to be run.
    Finally, the end user can use the class as a generator as the __next__ function exists.
    """

    # noinspection PyUnusedLocal
    def __init__(
        self,
        data: ByteString,
        result: str = None,  # pylint: disable=unused-argument
        offset: int = 0x0,
        collapse: bool = True,
    ):
        self._result = "".join(_line_gen(data, offset, collapse))
        self._result_lines = None
        self._line_pos = 0

    def __repr__(self):
        return self._result

    def __iter__(self):
        return self

    def __next__(self):
        if self._result_lines is None:
            # It's an expensive operation, delay until necessary
            self._result_lines = self._result.splitlines()

        try:
            line = self._result_lines[self._line_pos]
            self._line_pos += 1
        except IndexError as exp_index:
            # Allows for re-running the generator
            self._line_pos = 0
            raise StopIteration from exp_index
        return line
